der/derive/src/sequence.rs

```diff
@@ -50,27 +43,33 @@ impl DeriveSequence {
 
         Self {
             ident: input.ident,
-            lifetime,
+            generics: input.generics.clone(),
             fields,
         }
     }
 
     /// Lower the derived output into a [`TokenStream`].
     pub fn to_tokens(&self) -> TokenStream {
         let ident = &self.ident;
+        let mut generics = self.generics.clone();
 
-        let lifetime = match self.lifetime {
-            Some(ref lifetime) => quote!(#lifetime),
-            None => default_lifetime(),
-        };
-
-        // Lifetime parameters
-        // TODO(tarcieri): support multiple lifetimes
-        let lt_params = self
-            .lifetime
-            .as_ref()
-            .map(|_| lifetime.clone())
-            .unwrap_or_default();
+        // Use the first lifetime parameter as lifetime for Decode/Encode lifetime
+        // if none found, add one.
+        let lifetime = generics
+            .lifetimes()
+            .next()
+            .map(|lt| lt.lifetime.clone())
+            .unwrap_or_else(|| {
+                let lt = default_lifetime();
+                generics
+                    .params
+                    .insert(0, GenericParam::Lifetime(LifetimeParam::new(lt.clone())));
+                lt
+            });
+
+        // We may or may not have inserted a lifetime.
+        let (_impl_generics, ty_generics, where_clause) = self.generics.split_for_impl();
+        let (impl_generics, _ty_generics, _where_clause) = generics.split_for_impl();
```

Comment:
I think losing the same-named underscored variables might make it a bit clear what's coming from where:

```suggestion
        let (_, ty_generics, where_clause) = self.generics.split_for_impl();
        let (impl_generics, _, _) = generics.split_for_impl();
```